File: VCL/C++Builder/GeoCoding/MainForm.cpp
//=============================================================================
// This source code is a part of TatukGIS Developer Kernel.
//=============================================================================
//
//  GeoCoding Sample - C++Builder VCL
//  ===================================
//  Demonstrates address geocoding and shortest-path routing using TatukGIS DK.
//
//  Geocoding translates a human-readable address string (e.g. "Chrysler 1345")
//  into geographic coordinates by matching it against street-range fields stored
//  in a vector layer.  The TGIS_Geocoding class performs this matching entirely
//  on the client side using the loaded shapefile — no external service is needed
//  (unless OSM online routing is enabled).
//
//  Routing finds the shortest path between two geocoded points via TGIS_ShortestPath,
//  which builds a Dijkstra graph from the road network.  The path segments are
//  visualised by copying them into a separate in-memory "RouteDisplay" layer
//  rendered in red, with green/red point markers for start and end positions.
//
//  Key concepts shown:
//    - Opening a TatukGIS project (.ttkproject) and accessing a named layer
//    - Constructing TGIS_Geocoding with TIGER/Line address-range field mappings
//    - Using TGIS_Geocoding::Parse to resolve typed text to coordinates
//    - Building a TGIS_ShortestPath graph with per-link-type cost modifiers
//    - Handling the LinkTypeEvent callback to classify road segments by MTFCC code
//    - Drawing route results and address markers into a TGIS_LayerVector overlay
//
//  Check project\options\directories in case of any problems during compilation
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

//#include "GisLicense.hpp"  // Uncomment and configure if using a licensed build
#include "MainForm.h"
#include <typeinfo>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

/// <summary>
///   Classifies a road segment as highway (link type 0) or local road (link type 1).
///
///   The US Census MTFCC code "S1400" and higher identify local/residential
///   streets, service roads, and alleys.  Codes below "S1400" are interstates,
///   US routes, state routes, and other major roads.
///
///   The routing engine multiplies each segment's base cost by CostModifiers[type].
///   A slider value of 10 for highways (CostModifiers[0] ≈ 0) makes highways
///   essentially free, routing the path onto them preferentially.
/// </summary>
void __fastcall TForm1::doLinkType( TObject * _sender, TGIS_ShapeArc * _shape,
                         int &_type )
{
  if ( _shape->GetField( "MTFCC" ) >= "S1400" ) {
    // MTFCC >= "S1400" → local/residential road
    _type = 1 ;
  }
  else {
    // MTFCC < "S1400" → highway or major road
    _type = 0 ;
  }
}
//---------------------------------------------------------------------------

/// <summary>
///   Called when the form is created.
///   Opens the California street project, creates the route-display overlay
///   layer, and initialises the geocoding and routing engines.
/// </summary>
/// <remarks>
///   GIS->Lock() / GIS->Unlock() suppress intermediate redraws during batch
///   setup — the standard DK pattern for multi-step viewer modifications.
/// </remarks>
void __fastcall TForm1::FormCreate(TObject *Sender)
{
  GIS->Lock() ;

  // Open the TatukGIS project file which references California street data.
  GIS->Open( GisSamplesDataDirDownload() + "\\Samples\\Projects\\California.ttkproject" ) ;

  // Retrieve the streets layer by name so the geocoding and routing engines
  // can operate on it directly.
  layerSrc = (TGIS_LayerVector *)( GIS->Get( "streets" ) ) ;

  if (layerSrc == NULL) return ;
  if ( !layerSrc->InheritsFrom( __classid(TGIS_LayerVector) ) ) return ;

  // Set the initial view to the full extent of the street network.
  GIS->VisibleExtent = layerSrc->ProjectedExtent ;

  // ------------------------------------------------------------------
  // Create an in-memory overlay layer for route visualisation.
  // UseConfig = False prevents the DK from writing a style file.
  // Red line, 2-pixel width (negative = device pixels).
  // ------------------------------------------------------------------
  layerRoute = new TGIS_LayerVector ;
  layerRoute->UseConfig = False ;
  layerRoute->Params->Line->Color = TGIS_Color::Red ;
  layerRoute->Params->Line->Width = -2 ;
  layerRoute->Params->Marker->OutlineWidth = 1 ;
  layerRoute->Name = "RouteDisplay" ;
  layerRoute->CS = GIS->CS ;  // Must share the viewer's coordinate system
  GIS->Add( layerRoute ) ;

  // ------------------------------------------------------------------
  // Configure the geocoding engine with TIGER/Line field mappings.
  //   Offset    — distance along the segment to offset the geocoded point
  //               from the road node, avoiding exact node placement.
  //   RoadName  — attribute field containing the full street name.
  //   LFrom/LTo — address number range, left side of the street.
  //   RFrom/RTo — address number range, right side of the street.
  // ------------------------------------------------------------------
  geoObj = new TGIS_Geocoding( layerSrc ) ;
  geoObj->Offset   = 0.0001 ;
  geoObj->RoadName = "FULLNAME" ;
  geoObj->LFrom    = "LFROMADD" ;
  geoObj->LTo      = "LTOADD" ;
  geoObj->RFrom    = "RFROMADD" ;
  geoObj->RTo      = "RTOADD" ;

  costFactor = 1 ;  // Placeholder for future cost scaling

  // ------------------------------------------------------------------
  // Configure the routing engine.
  //   LoadTheData builds the Dijkstra graph from all arc shapes.
  //   LinkTypeEvent classifies each segment; RoadName labels route steps.
  // ------------------------------------------------------------------
  rtrObj = new TGIS_ShortestPath( *GIS ) ;
  rtrObj->LinkTypeEvent = doLinkType ;
  rtrObj->LoadTheData( layerSrc ) ;
  rtrObj->RoadName = "FULLNAME" ;

  GIS->Unlock() ;

  // Display the scale bar in miles (EPSG 9035).
  GIS_ControlScale1->Units = CSUnitsList()->ByEPSG( 9035 ) ;
}
//---------------------------------------------------------------------------

/// <summary>
///   Resolves the address in edtAddrFrom against the streets layer and
///   highlights all matching segments on the map.
///
///   If matches are found, the corresponding street shapes are copied to
///   layerRoute (red) and a green marker is placed at each geocoded point.
///   The view then zooms to the result set.  "???" is appended if no match
///   is found.
/// </summary>
void __fastcall TForm1::btnResolveClick(TObject *Sender)
{
  int i ;
  int r ;
  TGIS_Shape *shp ;

  if ( geoObj == NULL ) return ;

  // Clear any previously displayed route or address highlights.
  layerRoute->RevertShapes() ;

  // Parse returns the number of matching candidates; subtract 1 for the
  // last valid array index.  If no match, append "???" to indicate failure.
  r = geoObj->Parse( edtAddrFrom->Text, true, true ) - 1 ;
  if ( r <= 0 ) edtAddrFrom->Text = edtAddrFrom->Text + " ???" ;

  for (i = 0; i <= r; i++) {
    // Update the text box to show the canonical name of each matched street.
    edtAddrFrom->Text = geoObj->Query[i] ;
    Application->ProcessMessages() ;  // Allow the UI to refresh between results

    // Copy the primary matched segment to the route display layer.
    shp = layerSrc->GetShape( geoObj->Uid[i] ) ;
    layerRoute->AddShape( shp, False ) ;
    if (i == 0)
      layerRoute->Extent = shp->ProjectedExtent ;  // Seed the bounding extent

    // Add the secondary (adjacent/complementary) segment if available.
    shp = layerSrc->GetShape( geoObj->UidEx[i] ) ;
    if ( shp )
      layerRoute->AddShape( shp ) ;

    // Place a green marker at the geocoded point coordinate.
    // Lock/Unlock brackets shape modification to maintain internal extent.
    shp = layerRoute->CreateShape( TGIS_ShapeType::Point ) ;
    shp->Lock( TGIS_Lock::Extent ) ;
      shp->AddPart() ;
      shp->AddPoint( geoObj->Point[i] ) ;
      shp->Params->Marker->Color = TGIS_Color::Green ;
    shp->Unlock() ;
  }

  // Zoom to the matched results, pulling back slightly for context.
  GIS->Lock() ;
    GIS->VisibleExtent = layerRoute->ProjectedExtent ;
    GIS->Zoom = 0.7 * GIS->Zoom ;
  GIS->Unlock() ;
}
//---------------------------------------------------------------------------

/// <summary>
///   Geocodes both addresses, computes the shortest route between them,
///   and displays the result as a red path with turn-by-turn directions.
///
///   Cost modifiers derived from the highway/local-road sliders are applied
///   before routing so the user can tune road-type preference at runtime.
///
///   The Compass property encodes the turn direction: 0 = straight,
///   1..3 = right turns, 4 = U-turn, -1..-3 = left turns, -4 = left U-turn.
///   Consecutive steps on the same named street are merged in the directions.
/// </summary>
void __fastcall TForm1::btnRouteClick(TObject *Sender)
{
  int         i ;
  TGIS_Shape *shp ;
  int         res ;
  TGIS_Point  pt_a ;
  TGIS_Point  pt_b ;
  String      ang ;
  String      oldnam ;

  // Apply slider-derived cost modifiers.
  // CostModifiers[0] = highway cost; CostModifiers[1] = local-road cost.
  // Formula maps slider 1..10 to multiplier 1.0..~0.09 (lower = cheaper = preferred).
  rtrObj->CostModifiers[0] = 1.0 - 1.0/11 * trkHighways->Position ;
  rtrObj->CostModifiers[1] = 1.0 - 1.0/11 * trkSmallRoads->Position ;

  // Geocode the start address.
  res = geoObj->Parse( edtAddrFrom->Text ) ;
  if ( res > 0 ) edtAddrFrom->Text = geoObj->Query[0] ;
  else           edtAddrFrom->Text = edtAddrFrom->Text + " ???" ;

  if (res <= 0) return ;    // Abort if start address not resolved
  pt_a = geoObj->Point[0] ; // Geocoded start coordinate

  // Geocode the end address.
  res = geoObj->Parse( edtAddrTo->Text ) ;
  if (res > 0) edtAddrTo->Text = geoObj->Query[0] ;
  else         edtAddrTo->Text = edtAddrTo->Text + " ???" ;

  if (res <= 0) return ;    // Abort if end address not resolved
  pt_b = geoObj->Point[0] ; // Geocoded end coordinate

  // UpdateTheData refreshes the graph (e.g. after data edits).
  // Find computes the shortest path; Unproject converts viewer coords to layer coords.
  rtrObj->UpdateTheData() ;
  rtrObj->Find( layerRoute->Unproject( pt_a ), layerRoute->Unproject( pt_b ) ) ;

  // Build the turn-by-turn directions text.
  memRoute->Lines->BeginUpdate() ;
  memRoute->Clear() ;
  oldnam = "#$@3eqewe" ;  // Sentinel value to detect the first direction step

  for (i = 0; i <= rtrObj->ItemsCount - 1; i++) {
    // Convert numeric Compass value to a direction abbreviation.
    switch ( rtrObj->Items[i]->Compass ) {
      case  0 : ang = "FWD  " ; break ;
      case  1 : ang = "RIGHT" ; break ;  // Slight right
      case  2 : ang = "RIGHT" ; break ;  // Right
      case  3 : ang = "RIGHT" ; break ;  // Sharp right
      case  4 : ang = "BACK " ; break ;  // U-turn
      case -1 : ang = "LEFT " ; break ;  // Slight left
      case -2 : ang = "LEFT " ; break ;  // Left
      case -3 : ang = "LEFT " ; break ;  // Sharp left
      case -4 : ang = "BACK " ; break ;  // Left U-turn
    }

    // Skip consecutive segments with the same street name to reduce clutter.
    if ( oldnam == rtrObj->Items[i]->Name ) continue ;
    oldnam = rtrObj->Items[i]->Name ;

    memRoute->Lines->Add( ang + " " + rtrObj->Items[i]->Name ) ;
  }
  memRoute->Lines->EndUpdate() ;

  // Repopulate the route display layer with the computed path segments.
  layerRoute->RevertShapes() ;

  for (i = 0; i <= rtrObj->ItemsCount - 1; i++) {
    shp = rtrObj->Items[i]->Layer->GetShape( rtrObj->Items[i]->Uid ) ;
    if ( shp == NULL ) continue ;
    layerRoute->AddShape( shp, False ) ;
    if (i == 0)
      layerRoute->Extent = shp->Extent ;  // Seed the bounding extent
  }

  // Place a green marker at the start point.
  shp = layerRoute->CreateShape( TGIS_ShapeType::Point ) ;
  shp->Lock( TGIS_Lock::Extent ) ;
    shp->AddPart() ;
    shp->AddPoint( pt_a ) ;
    shp->Params->Marker->Color = TGIS_Color::Green ;
  shp->Unlock() ;

  // Place a red (default colour) marker at the end point.
  shp = layerRoute->CreateShape( TGIS_ShapeType::Point ) ;
  shp->Lock( TGIS_Lock::Extent ) ;
    shp->AddPart() ;
    shp->AddPoint( pt_b ) ;
  shp->Unlock() ;

  // Zoom to the route bounding box, pulling back slightly for context.
  GIS->Lock() ;
    GIS->VisibleExtent = layerRoute->Extent ;
    GIS->Zoom = 0.7 * GIS->Zoom ;
  GIS->Unlock() ;
}
//---------------------------------------------------------------------------

/// <summary>
///   Frees the geocoding and routing engine objects when the form is destroyed.
///   layerRoute is owned by the viewer (added via GIS->Add) and is freed
///   automatically when the viewer is destroyed.
/// </summary>
void __fastcall TForm1::FormDestroy(TObject *Sender)
{
  delete geoObj ;
  delete rtrObj ;
}
//---------------------------------------------------------------------------

File: VCL/Delphi/GeoCoding/MainForm.pas
//=============================================================================
// This source code is a part of TatukGIS Developer Kernel.
//=============================================================================
{
  GeoCoding Sample - Delphi VCL
  ==============================
  Demonstrates address geocoding and shortest-path routing using TatukGIS DK11.

  Geocoding translates a human-readable address string (e.g. "Chrysler 1345")
  into geographic coordinates by matching it against street-range fields stored
  in a vector layer.  The TGIS_Geocoding class performs this matching entirely
  on the client side using a loaded shapefile — no external service is required
  unless OSM online routing is enabled.

  Routing then finds the shortest path between two geocoded points using
  TGIS_ShortestPath, which builds an in-memory graph from the road network and
  applies Dijkstra's algorithm.  The resulting path segments are visualised by
  adding them to a separate in-memory "RouteDisplay" layer rendered in red.

  Key concepts shown:
    - Loading a TatukGIS project file (.ttkproject) containing a roads layer
    - Constructing TGIS_Geocoding with address-range field mappings
      (FULLNAME, LFROMADD/LTOADD for left side, RFROMADD/RTOADD for right side)
    - Using TGIS_Geocoding.Parse to resolve a typed address to coordinates
    - Building a TGIS_ShortestPath graph with per-link-type cost modifiers
    - Handling the LinkTypeEvent callback to classify road segments
    - Drawing route results into an in-memory TGIS_LayerVector
    - Optionally switching to OSM online geocoding/routing via chkbxOnline

  Data used:
    California.ttkproject — a TatukGIS project referencing US Census TIGER/Line
    street data for California.  Field layout:
      FULLNAME   — full street name
      LFROMADD   — address range start, left side of street
      LTOADD     — address range end, left side of street
      RFROMADD   — address range start, right side of street
      RTOADD     — address range end, right side of street
      MTFCC      — MAF/TIGER Feature Class Code; "S1400" = local/residential road
}

unit MainForm;

interface

uses
  System.Classes,
  System.SysUtils,

  Vcl.Forms,
  Vcl.ComCtrls,
  Vcl.StdCtrls,
  Vcl.ToolWin,
  Vcl.Controls,
  Vcl.Graphics,
  Vcl.ImgList,
  Vcl.ExtCtrls,

  //GisLicense,       // Uncomment and configure if using a licensed build
  GisTypes,           // Core TatukGIS types (TGIS_Point, extents, shape types, …)
  GisCsBase,          // Coordinate-system base classes
  GisEditor,          // Layer editing support
  GisGeocoding,       // TGIS_Geocoding — address-to-coordinate resolution
  GisLayerVector,     // TGIS_LayerVector — in-memory and file-backed vector layers
  GisLayerSHP,        // Shapefile driver (registers .shp support)
  GisLayerADF,        // ArcInfo Binary Grid driver (used by some project files)
  GisParams,          // Rendering parameter classes (colour, width, markers, …)
  GisShortestPath,    // TGIS_ShortestPath — Dijkstra graph-based routing engine
  GisTopology,        // Topology utilities used internally by the routing engine
  GisTypesUI,         // TGIS_ViewerMode, TGIS_Color, and other UI-facing types
  GisUtils,           // GisSamplesDataDirDownload, CSUnitsList, and helpers
  GisViewer,          // Base viewer class

  Vcl.GisControlScale, // Scale-bar overlay control
  Vcl.GisViewerWnd ;   // VCL windowed map viewer component

type
  /// <summary>
  ///   Main application form for the GeoCoding sample.
  ///   Hosts a TGIS_ViewerWnd showing California street data and provides
  ///   address resolution (geocoding) and route finding (shortest path).
  /// </summary>
  TForm1 = class(TForm)
    GIS: TGIS_ViewerWnd;                 // Map viewer — renders street data and the route
    GIS_ControlScale1: TGIS_ControlScale; // Scale-bar overlay linked to the viewer
    Panel2: TPanel;                       // Right-hand panel hosting the controls
    memRoute: TMemo;                      // Read-only text area showing turn-by-turn directions
    GroupBox1: TGroupBox;                 // Groups the routing parameter controls
    trkSmallRoads: TTrackBar;            // Cost modifier slider: preference for local roads (1–10)
    trkHighways: TTrackBar;              // Cost modifier slider: preference for highways (1–10)
    lblSmallRoads: TLabel;               // Label for trkSmallRoads
    lblHighways: TLabel;                 // Label for trkHighways
    edtAddrFrom: TEdit;                  // Start address text box (e.g. "Chrys 1345")
    btnResolve: TButton;                 // Resolve and highlight the start address on the map
    edtAddrTo: TEdit;                    // End address text box (e.g. "wash")
    btnRoute: TButton;                   // Geocode both addresses and find the shortest route
    lblAddrFrom: TLabel;                 // Label for edtAddrFrom
    lblAddrTo: TLabel;                   // Label for edtAddrTo
    chkbxOnline: TCheckBox;             // Toggle between local-data and OSM online geocoding/routing

    procedure FormCreate(Sender: TObject);
    procedure btnResolveClick(Sender: TObject);
    procedure btnRouteClick(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure chkbxOnlineClick(Sender: TObject);
  private
    layerSrc   : TGIS_LayerVector ;  // Source streets layer loaded from the project file
    layerRoute : TGIS_LayerVector ;  // In-memory layer used to display the computed route
    rtrObj     : TGIS_ShortestPath ; // Routing engine; builds graph from layerSrc
    geoObj     : TGIS_Geocoding ;    // Geocoding engine; resolves address text to coordinates
    costFactor : Double ;            // Reserved cost scaling factor (currently fixed at 1)

    /// <summary>
    ///   Callback invoked by TGIS_ShortestPath for each road segment to classify
    ///   its link type.  Link type 0 = highway; link type 1 = local road.
    ///   The router uses these types together with CostModifiers to weigh edges.
    /// </summary>
    /// <param name="_shape">
    ///   The arc shape (road segment) being classified.
    /// </param>
    /// <param name="_type">
    ///   Out parameter: set to 1 for local roads (MTFCC >= "S1400"), 0 for
    ///   highways and major roads.
    /// </param>
    procedure doLinkType(     _sender   : TObject ;
                              _shape    : TGIS_ShapeArc ;
                          var _type     : Integer
                        ) ;
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}


/// <summary>
///   Classifies a road segment as highway (0) or local road (1).
///
///   The US Census MTFCC code "S1400" and higher identify local residential
///   streets, service roads, and alleys.  Codes below "S1400" are major roads
///   (interstates, US routes, state routes, etc.).
///
///   The cost modifiers applied via CostModifiers[0] and CostModifiers[1]
///   scale the traversal cost for each link type.  A lower cost makes the
///   router prefer that type; sliders let the user tune this at runtime.
/// </summary>
procedure TForm1.doLinkType(    _sender   : TObject ;
                                 _shape    : TGIS_ShapeArc ;
                             var _type     : Integer
                           ) ;
begin
  if _shape.GetField( 'MTFCC' ) >= 'S1400' then
    // MTFCC >= S1400 → local/residential road (link type 1)
    _type := 1
  else
    // MTFCC < S1400 → highway or major road (link type 0)
    _type := 0 ;
end ;

/// <summary>
///   Called when the form is created.
///   Opens the California street project, builds the geocoding and routing
///   objects, and sets up the route-display layer.
/// </summary>
/// <remarks>
///   The viewer is locked during setup to suppress intermediate redraws.
///   GIS.Lock / GIS.Unlock are the standard DK pattern for batch updates.
/// </remarks>
procedure TForm1.FormCreate(Sender: TObject);
begin
  GIS.Lock ;

  // Open the TatukGIS project file which references the California street data.
  // The project loads the "streets" layer automatically.
  GIS.Open( TGIS_Utils.GisSamplesDataDirDownload + 'Samples\Projects\California.ttkproject' ) ;

  // Retrieve the streets layer by name so we can pass it to the geocoding
  // and routing engines.  Cast is safe because the project was built with
  // a vector layer at that name.
  layerSrc := TGIS_LayerVector( GIS.Get( 'streets' ) ) ;

  if not Assigned( layerSrc ) then exit ;
  if not ( layerSrc is TGIS_LayerVector ) then exit ;

  // Zoom to the full extent of the street network on startup.
  GIS.VisibleExtent := layerSrc.ProjectedExtent ;

  // ------------------------------------------------------------------
  // Create an in-memory vector layer to display the computed route.
  // UseConfig = False prevents the DK from writing a style config file.
  // The line is rendered in red with a 2-pixel width.
  // Markers (for start/end points) will use a 1-pixel outline.
  // ------------------------------------------------------------------
  layerRoute := TGIS_LayerVector.Create ;
  layerRoute.UseConfig := False ;
  layerRoute.Params.Line.Color := TGIS_Color.Red ;
  layerRoute.Params.Line.Width := -2 ;             // Negative = device pixels (DPI-independent)
  layerRoute.Params.Marker.OutlineWidth := 1 ;
  layerRoute.Name := 'RouteDisplay' ;
  layerRoute.CS := GIS.CS ;                        // Must share the viewer's coordinate system
  GIS.Add( layerRoute ) ;

  // ------------------------------------------------------------------
  // Configure the geocoding engine.
  //   Offset    — how far along the road segment to place the geocoded
  //               point (in layer units), avoiding placing it on the
  //               exact road node.
  //   RoadName  — attribute field containing the full street name.
  //   LFrom/LTo — address number range for the left side of the street.
  //   RFrom/RTo — address number range for the right side of the street.
  //               US Census TIGER/Line uses these four fields to
  //               interpolate the house position along the segment.
  // ------------------------------------------------------------------
  geoObj := TGIS_Geocoding.Create( layerSrc ) ;
  geoObj.offset    := 0.0001 ;
  geoObj.RoadName  := 'FULLNAME' ;
  geoObj.LFrom     := 'LFROMADD' ;
  geoObj.LTo       := 'LTOADD' ;
  geoObj.RFrom     := 'RFROMADD' ;
  geoObj.RTo       := 'RTOADD' ;

  costFactor := 1 ;   // Placeholder for future cost scaling

  // ------------------------------------------------------------------
  // Configure the routing engine.
  //   LoadTheData builds the in-memory graph from all arc shapes in
  //   layerSrc.  LinkTypeEvent is called for each segment to assign a
  //   link type, enabling per-type cost modifiers.
  //   RoadName is used to label each route step in the directions panel.
  // ------------------------------------------------------------------
  rtrObj := TGIS_ShortestPath.Create( GIS ) ;
  rtrObj.LinkTypeEvent := doLinkType ;
  rtrObj.LoadTheData( layerSrc ) ;
  rtrObj.RoadName := 'FULLNAME' ;

  GIS.Unlock ;

  // Display the scale bar in miles (EPSG 9035).
  GIS_ControlScale1.Units := CSUnitsList.ByEPSG( 9035 ) ;
end ;

/// <summary>
///   Resolves the address typed in edtAddrFrom against the streets layer
///   and highlights all matching segments on the map.
///
///   If one or more matches are found, the matching street shapes are copied
///   to layerRoute (shown in red) and a green marker is placed at the exact
///   geocoded point for each result.  The view zooms to the result set.
///
///   If no matches are found, " ???" is appended to the text box to indicate
///   the address could not be resolved.
/// </summary>
procedure TForm1.btnResolveClick(Sender: TObject);
var
  i   : Integer ;
  r   : Integer ;
  shp : TGIS_Shape ;
begin
  if not Assigned( geoObj ) then exit ;

  // Clear any previously displayed route or address result.
  layerRoute.RevertShapes ;

  // Parse the address string.  Returns the number of matching candidates.
  // Subtracting 1 gives the last valid index for the result arrays.
  r := geoObj.Parse( edtAddrFrom.Text ) - 1 ;
  if r <= 0 then edtAddrFrom.Text := edtAddrFrom.Text + ' ???' ;

  for i := 0 to r do
  begin
    // Show the canonical name of each matching street in the text box.
    edtAddrFrom.Text := geoObj.Query[i] ;
    Application.ProcessMessages ; // Allow the UI to refresh between results

    // Add the primary matching street segment to the route display layer.
    // GetShape retrieves the shape by its unique ID from the source layer.
    shp := layerSrc.GetShape( geoObj.Uid[i] ) ;
    layerRoute.AddShape( shp ) ;

    // Initialise the display extent from the first matched segment.
    if i = 0 then layerRoute.Extent := shp.ProjectedExtent ;

    // Add the secondary (adjacent) segment if available.
    // UidEx returns the ID of the complementary segment for the same address
    // range (e.g. the opposite side of the street).
    shp := layerSrc.GetShape( geoObj.UidEx[i] ) ;
    if Assigned( shp ) then
      layerRoute.AddShape( shp ) ;

    // Place a green point marker at the exact geocoded coordinate.
    // Lock/Unlock brackets shape modification to keep internal extent correct.
    shp := layerRoute.CreateShape( TGIS_ShapeType.Point ) ;
    shp.Lock( TGIS_Lock.Extent ) ;
    shp.AddPart ;
    // FromCS converts the point from the source layer's CS to the route layer's CS.
    shp.AddPoint( layerRoute.CS.FromCS( layerSrc.CS, geoObj.Point[i] ) ) ;
    shp.Params.Marker.Color := TGIS_Color.Green ;
    shp.Unlock ;
  end ;

  // Zoom to the matched result with a slight zoom-out factor (0.7×) so the
  // surrounding context is also visible.
  GIS.Lock ;
  GIS.VisibleExtent := layerRoute.ProjectedExtent ;
  GIS.Zoom := 0.7 * GIS.Zoom ;
  GIS.Unlock ;
end ;

/// <summary>
///   Geocodes both the From and To addresses, then computes and displays the
///   shortest route between them.
///
///   The turn-by-turn directions are written to memRoute as compass direction
///   abbreviations (FWD, LEFT, RIGHT, BACK) followed by the street name.
///   Route segments are added to layerRoute (red lines) and the start and
///   end points are marked with green and red markers respectively.
/// </summary>
/// <remarks>
///   Cost modifiers are re-applied before each routing call so the user can
///   adjust the highway/local-road preference without reloading data.
///
///   The Compass property of each route item encodes the turn direction as an
///   integer: 0 = straight, 1..3 = right turns (45°, 90°, 135°),
///   4 = U-turn, -1..-3 = left turns, -4 = left U-turn.
/// </remarks>
procedure TForm1.btnRouteClick(Sender: TObject);
var
  i      : Integer ;
  shp    : TGIS_Shape ;
  res    : Integer ;
  pt_a   : TGIS_Point ;
  pt_b   : TGIS_Point ;
  ang    : String ;
  oldnam : String ;
begin
  // Apply slider-derived cost modifiers.
  // CostModifiers[0] scales highway cost; CostModifiers[1] scales local-road cost.
  // Formula maps slider value 1..10 to modifier 1.0..~0.09 (lower = cheaper = preferred).
  rtrObj.CostModifiers[0] := 1 - 1/11 * trkHighways.Position ;
  rtrObj.CostModifiers[1] := 1 - 1/11 * trkSmallRoads.Position ;

  // Geocode the start address.  Parse returns the number of candidates found.
  // If any candidates are found (res > 0), refine the text box to the best match.
  res := geoObj.Parse( edtAddrFrom.Text ) ;
  if res > 0 then edtAddrFrom.Text := geoObj.Query[0]
             else edtAddrFrom.Text := edtAddrFrom.Text + ' ???' ;

  // Abort if the start address could not be resolved.
  if res <= 0 then exit ;
  pt_a := geoObj.Point[0] ;  // Geocoded start coordinate

  // Geocode the end address using the same approach.
  res := geoObj.Parse( edtAddrTo.Text ) ;
  if res > 0 then edtAddrTo.Text := geoObj.Query[0]
             else edtAddrTo.Text := edtAddrTo.Text + ' ???' ;

  // Abort if the end address could not be resolved.
  if res <= 0 then exit ;
  pt_b := geoObj.Point[0] ;  // Geocoded end coordinate

  // UpdateTheData refreshes the routing graph (e.g. after the layer was edited).
  // Find computes the shortest path between the two projected points.
  // Unproject converts from the viewer's display CS back to layer coordinates.
  rtrObj.UpdateTheData ;
  rtrObj.Find( layerRoute.Unproject( pt_a ),
               layerRoute.Unproject( pt_b ) ) ;

  // Build the turn-by-turn directions text.
  // BeginUpdate/EndUpdate prevents excessive redraws while populating the memo.
  memRoute.Lines.BeginUpdate ;
  memRoute.Clear ;
  oldnam := '#$@3eqewe' ;  // Sentinel to detect the first segment

  for i := 0 to rtrObj.ItemsCount - 1 do
  begin
    // Map the numeric Compass value to a human-readable direction string.
    case rtrObj.Items[i].Compass of
      0  : ang := 'FWD  ' ;
      1  : ang := 'RIGHT' ;  // Slight right
      2  : ang := 'RIGHT' ;  // Right
      3  : ang := 'RIGHT' ;  // Sharp right
      4  : ang := 'BACK ' ;  // U-turn
      -1 : ang := 'LEFT ' ;  // Slight left
      -2 : ang := 'LEFT ' ;  // Left
      -3 : ang := 'LEFT ' ;  // Sharp left
      -4 : ang := 'BACK ' ;  // Left U-turn
    end ;

    // Suppress consecutive items on the same named street to avoid clutter.
    if oldnam = rtrObj.Items[i].Name then continue ;
    oldnam := rtrObj.Items[i].Name ;

    memRoute.Lines.Add( Format( '%s %s', [ ang, rtrObj.Items[i].Name ] ) ) ;
  end ;
  memRoute.Lines.EndUpdate ;

  // Clear the route display layer and repopulate with the new path segments.
  layerRoute.RevertShapes ;

  // Copy each route segment shape from the source layer into the display layer.
  for i := 0 to rtrObj.ItemsCount - 1 do
  begin
    shp := rtrObj.Items[i].Layer.GetShape( rtrObj.Items[i].Uid ) ;
    if not Assigned( shp ) then continue ;
    layerRoute.AddShape( shp ) ;
    if i = 0 then
      layerRoute.Extent := shp.Extent ;  // Seed the bounding extent from the first shape
  end ;

  // Place a green point marker at the start (pt_a).
  shp := layerRoute.CreateShape( TGIS_ShapeType.Point ) ;
  shp.Lock( TGIS_Lock.Extent ) ;
  shp.AddPart ;
  shp.AddPoint( pt_a ) ;
  shp.Params.Marker.Color := TGIS_Color.Green ;
  shp.Unlock ;

  // Place a red (default) point marker at the end (pt_b).
  shp := layerRoute.CreateShape( TGIS_ShapeType.Point ) ;
  shp.Lock( TGIS_Lock.Extent ) ;
  shp.AddPart ;
  shp.AddPoint( pt_b ) ;
  shp.Unlock ;

  // Zoom to the route bounding box, pulling back slightly for context.
  GIS.Lock ;
  GIS.VisibleExtent := layerRoute.Extent ;
  GIS.Zoom := 0.7 * GIS.Zoom ;
  GIS.Unlock ;
end ;

/// <summary>
///   Toggles between local-data geocoding/routing and OSM online services.
///
///   When OSMGeocoding is true, TGIS_Geocoding queries the OpenStreetMap
///   Nominatim service instead of the local layer data.  When OSMRouting is
///   true, TGIS_ShortestPath uses the OSRM routing service.
///   Both services require an internet connection.
/// </summary>
procedure TForm1.chkbxOnlineClick(Sender: TObject);
begin
  if chkbxOnline.Checked then begin
    geoObj.OSMGeocoding := True ;
    rtrObj.OSMRouting   := True ;
  end
  else begin
    geoObj.OSMGeocoding := False ;
    rtrObj.OSMRouting   := False ;
  end ;
end ;

/// <summary>
///   Frees the geocoding and routing objects when the form is destroyed.
///   layerRoute is owned by the viewer and is freed automatically.
/// </summary>
procedure TForm1.FormDestroy(Sender: TObject);
begin
  geoObj.Free ;
  rtrObj.Free ;
end ;

end.
